readium/shared/src/main/java/org/readium/r2/shared/publication/Href.kt

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2020 Readium Foundation. All rights reserved.
+ * Use of this source code is governed by the BSD-style license
+ * available in the top-level LICENSE file of the project.
+ */
+
+package org.readium.r2.shared.publication
+
+import android.os.Parcelable
+import kotlinx.parcelize.IgnoredOnParcel
+import kotlinx.parcelize.Parcelize
+import org.readium.r2.shared.ExperimentalReadiumApi
+import org.readium.r2.shared.util.URITemplate
+import org.readium.r2.shared.util.Url as SharedUrl
+import org.readium.r2.shared.util.toUrl
+import timber.log.Timber
+
+/**
+ * An hypertext reference points to a resource in a [Publication].
+ */
+@Parcelize
+public class Href private constructor(private val href: Url) : Parcelable {
+
+    public companion object {
+        public operator fun invoke(href: SharedUrl): Href =
+            Href(StaticUrl(href))
+
+        public operator fun invoke(href: String, templated: Boolean = false): Href? {
+            val url =
+                if (templated) {
+                    TemplatedUrl(href)
+                } else {
+                    SharedUrl(href)?.let { StaticUrl(it) }
+                }
+
+            return url?.let { Href(it) }
+        }
+    }
+
+    /**
+     * Returns the URL represented by this HREF, resolved to the given [base] URL.
+     *
+     * If the HREF is a template, the [parameters] are used to expand it according to RFC 6570.
+     */
+    public fun toUrl(
+        base: SharedUrl? = null,
+        parameters: Map<String, String> = emptyMap()
+    ): SharedUrl = href.toUrl(base, parameters)
+
+    /**
+     * Syntactic sugar for [toUrl].
+     */
+    public operator fun invoke(
+        base: SharedUrl? = null,
+        parameters: Map<String, String> = emptyMap()
+    ): SharedUrl = href.toUrl(base, parameters)
```

Comment:
This makes some sense if you see the `href` as a function. In practice I find it a bit confusing when it is used as in `link.href()` though. It would be clearer to me to have both the `href: Href` property which would denote some data (it's a JSON model)  and a `url(base, parameters)` method which would resolve the href to an URL.